content/6_Advanced/LineContainer.mdx

```diff
@@ -30,13 +30,201 @@ frequency: 1
+};
+
+// Cross product of two vectors
+long double cross(const Point &a, const Point &b) {
+	return a.x * b.y - a.y * b.x;
+}
+
+struct Halfplane {
+	Point p, pq;
+	long double angle;
+
+	Halfplane() {}
+	Halfplane(const Point &a, const Point &b) : p(a), pq(b - a) {
+		angle = atan2l(pq.y, pq.x);  // Compute angle of direction vector
+	}
+
+	// Check if a point r is outside the half-plane
+	bool out(const Point &r) const { return cross(pq, r - p) < -eps; }
+
+	// Comparator for sorting half-planes by angle
+	bool operator<(const Halfplane &rhs) const { return angle < rhs.angle; }
+};
+
+// Function to compute the intersection of a set of half-planes
+std::vector<Point> halfplane_intersect(std::vector<Halfplane> &halfplanes) {
+	// Define a bounding box as 4 half-planes
+	Point box[4] = {Point(inf, inf), Point(-inf, inf), Point(-inf, -inf),
+	                Point(inf, -inf)};
+
+	// Add bounding box halfplanes to the input set
+	for (int i = 0; i < 4; i++) {
+		Halfplane aux(box[i], box[(i + 1) % 4]);
+		halfplanes.push_back(aux);
+	}
+
+	// Sort half-planes by angle (counter-clockwise)
+	std::sort(halfplanes.begin(), halfplanes.end());
+
+	// Remove duplicates (half-planes with same direction)
+	halfplanes.erase(std::unique(halfplanes.begin(), halfplanes.end()),
+	                 halfplanes.end());
+
+	std::deque<Halfplane> active_halfplanes;
+	int len = 0;
+
+	// Process each sorted half-plane
+	for (int i = 0; i < int(halfplanes.size()); i++) {
+		// Remove invalid half-planes from the end of active set
+		while (len > 1 &&
+		       halfplanes[i].out(intersection(active_halfplanes[len - 1],
+		                                      active_halfplanes[len - 2]))) {
+			active_halfplanes.pop_back();
+			len--;
+		}
+
+		// Remove invalid half-planes from the beginning of active set
+		while (len > 1 && halfplanes[i].out(intersection(
+		                      active_halfplanes[0], active_halfplanes[1]))) {
+			active_halfplanes.pop_front();
+			len--;
+		}
+
+		// Add current half-plane to active set
+		active_halfplanes.push_back(halfplanes[i]);
+		len++;
+	}
+
+	// Check if there are less than 3 valid half-planes
+	if (len < 3) return std::vector<Point>();
+
+	// Compute intersection points of active half-planes
+	std::vector<Point> intersection_points(len);
+	for (int i = 0; i < len - 1; i++) {
+		intersection_points[i] =
+		    intersection(active_halfplanes[i], active_halfplanes[i + 1]);
+	}
+	intersection_points.back() =
+	    intersection(active_halfplanes[len - 1], active_halfplanes[0]);
+
+	return intersection_points;
+}
+
+// Function to compute intersection of two half-planes
+Point intersection(const Halfplane &h1, const Halfplane &h2) {
+	long double den = cross(h1.pq, h2.pq);
+	long double num1 = cross(h2.pq, h1.p - h2.p);
+
+	// Calculate intersection point
+	return Point(h1.p.x + (num1 / den) * h1.pq.x,
+	             h1.p.y + (num1 / den) * h1.pq.y);
+}
+
+int main() {
+	std::vector<Halfplane> halfplanes;
```

Comment:
bruh
what problem does this ac on

Reply:
the example one

Reply:
no most modules give an algorithm _with_ a cp-style problem

Reply:
so, now should i add that problem and its solution in the module as well (the one i submitted proof of ac for)?

Reply:
but the solution still outputs what the algorithm is supposed to do; it gives the intersection points. so a cp problem doesn't matter??? the problem explanation in the module is still a great problem.

Reply:
the whole point of usaco guide is to give an actual problem to apply these algorithms to in both examples and exercises

Reply:
so should i add practice problems

Reply:
yes
i thought this was well established 💀

Reply:
and should i replace that ac problem with its new solution in place of the existing one? or should i add below the existing explanation and all?

Reply:
replace

Reply:
oh btw can i see this new problem? it has to be an extremely straightforward application of the algorithm you describe- just like the other cses problems the guide showcases

Reply:
yeah, the same one i got ac for:
https://www.codechef.com/problems/CHN02

Reply:
also idk how to add another file to this pr, so can i add that focus problem and other practice problems in the json file in a separate pr? as in reopen that closed pr...

Reply:
i
how do you not know
how to commit multiple files

Reply:
but yeah the problem seems nice

Reply:
ok i'll figure it out.

Reply:
it automatically makes a different pr, i tried. so can you reopen that closed pr? then, i can also add the new ac cp-style focus problem, in addition to the practice problems i already committed there.

Reply:
oh.
are you working on the usaco guide editor?
please just...git clone it locally

Reply:
no, i'm working on the web editor on github.
idk does the same thing?
so reopen that pr, things will be much easier and faster?

Reply:
no it's very limiting
just git clone

Reply:
ok
edit: i tried but nope it doesn't work, still requires some merging that creates a diff pr.
so rather i'll make another pr for the json file myself, like before. i don't see how it's limiting; moreover, it's more flexible.

Reply:
i have never heard of anyone having to make two separate prs for two separate files
are you sure you're on the right branch?

Reply:
yeah

Reply:
yeah
so, i'm merging that closed pr's branch of added problems into this pr's branch in my forked repository, but it still makes a diff pr for merging that into this. so, rather i'll make a diff pr myself to separately update it.
bcz this is taking much longer than i expected, like so close to a 100 conversations, i've never seen that number in any pr.
reopening that is much easier and definitely faster.

Reply:
> can you tell me what exactly you're doing, in terms of commits, pushes, whatever?
maybe a transcript of what shows on your terminal would help as well

Reply:
uh, i just did?

Reply:
i'll add the ac cp-style problem now

Reply:
update: i added the focus problem

Reply:
where is it

Reply:
right after the resources and before the explanation

Reply:
i added it as a focus problem like in other modules, followed by explanation and solution.

Reply:
i also added the practice problems

Reply:
the main method doesn't seem to read any input from stdin

Reply:
i modified it. now it should work.

Reply:
@SansPapyrus683 is there anything else to be reviewed and modified?

Reply:
uh overall it looks fine, but the explanation needs to be looked at and i'm a bit busy lol
sorry for the inconvenience

Reply:
oh it's ok